in-progress/0003-native-merkle-trees.md

```diff
@@ -0,0 +1,212 @@
+    getSnapshot(block: number): Promise<TreeSnapshots>;
+  };
+```
+
+An abstract factory will then be created to construct the appropriate concrete type based on whether an instance of the node has native World State or not.
+
+### Interface
+
+The interface will be an asynchronous message based communication protocol. Each message is provided with meta data uniquely identiying it and is responded to inidividually. It is not necessary to wait for a response to a message before sending a subsequent message. A simple message specification will be created, some examples of which are shown here:
+
+``` C++
+enum WorldStateMsgTypes {
+    START_TREE_REQUEST = FIRST_APP_MSG_TYPE,
+    START_TREE_RESPONSE,
+    GET_TREE_INFO_REQUEST,
+    GET_TREE_INFO_RESPONSE,
+    INSERT_LEAVES_REQUEST,
+    INSERT_LEAVES_RESPONSE,
+};
+
+struct MsgHeader {
+    uint32_t messageId; // Unique Id for the message
+    uint32_t requestId; // Id of the message this is responding too (may not be used)
+
+    MSGPACK_FIELDS(messageId, requestId);
+
+    MsgHeader() = default;
+
+    MsgHeader(uint32_t reqId)
+        : requestId(reqId)
+    {}
+
+    MsgHeader(uint32_t msgId, uint32_t reqId)
+        : messageId(msgId)
+        , requestId(reqId)
+    {}
+};
+
+struct GetTreeInfoRequest {
+    std::string name;
+
+    MSGPACK_FIELDS(name);
+};
+
+struct GetTreeInfoResponse {
+    std::string name;
+    uint32_t depth;
+    bb::fr root;
+    uint64_t size;
+    bool success;
+    std::string message;
+
+    MSGPACK_FIELDS(name, depth, root, size, success, message);
+};
+
+template <class T> struct TypedMessage {
+    uint32_t msgType;
+    MsgHeader header;
+    T value;
+
+    TypedMessage(uint32_t type, MsgHeader& hdr, const T& val)
+        : msgType(type)
+        , header(hdr)
+        , value(val)
+    {}
+
+    TypedMessage() = default;
+
+    MSGPACK_FIELDS(msgType, header, value);
+};
+```
+
+``` TS
+export type GetTreeInfoRequest = {
+  name: string;
+}
+
+export type GetTreeInfoResponse = {
+  name: string;
+  depth: number;
+  success: boolean;
+  message: string;
+  root: Buffer;
+  size: bigint;
+}
+```
+
+### LMDB
+
+LMDB is a high performance key-value database allowing for concurrent read/write access and fully ACID transactions. In particular, it supports up to 126 concurrent read transactions. Write transactions can be performed concurrently with reads but we won't use this. The majority of our World State operations only require read access to persisted data.
+
+There are 3 broad categories of World State operations:
+
+#### Reads
+
+Simply reading data from the trees is performed using either `committed` or `uncommitted` state. Committed state is that which has fully settled and is therefore not going to change over the course of building a block. It can only change upon settlement of a new block. Uncommitted reads will read from the pending state, it is not recommended that uncommitted reads are performed by anyone other than a sequencer/prover.
```

Comment:
> settlement of a new block

This "block" is a miniblock/batch right? So it is conceivable that the `committed` state can get wiped in the event of an L2 reorg?

Reply:
This work only really considers the current model of synching a block once it's settled. Not the upcoming work to introduce batches.

With batches we may need to have an additional layer of state. So e.g.

1. Data becomes 'committed' once it has been fully proven.
2. Data is 'checkpointed' once it is in a batch.
3. Data is 'uncommitted' when a sequencer has inserted it into the tree as part of proposing a batch.

Type 3 can be rolled back to 2. And 2 can be rolled back to 1.